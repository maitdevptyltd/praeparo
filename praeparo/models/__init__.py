"""Pydantic models describing Praeparo configuration objects."""

from .matrix import MatrixConfig, MatrixTotals, MatrixValueConfig, RowTemplate

__all__ = ["MatrixConfig", "MatrixTotals", "MatrixValueConfig", "RowTemplate"]
