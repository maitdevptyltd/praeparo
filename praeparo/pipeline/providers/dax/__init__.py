"""DAX execution client abstractions."""

from .clients import DaxExecutionClient, PowerBIDaxClient

__all__ = [
    "DaxExecutionClient",
    "PowerBIDaxClient",
]
