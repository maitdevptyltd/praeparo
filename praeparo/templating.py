"""Template utilities for extracting field references."""

from __future__ import annotations

from collections import OrderedDict
from dataclasses import dataclass
import re
from typing import Iterable, Iterator

JINJA_PLACEHOLDER = re.compile(r"\{\{\s*(?P<expr>[^}]+?)\s*\}\}")


@dataclass(frozen=True)
class FieldReference:
    """Represents a data field referenced within a Jinja placeholder."""

    expression: str
    table: str | None
    column: str

    @property
    def dax_reference(self) -> str:
        """Return the DAX column reference for this field."""

        if self.table:
            return f"{self.table}[{self.column}]"
        return f"[{self.column}]"

    @property
    def placeholder(self) -> str:
        """Return the canonical placeholder expression."""

        if self.table:
            return f"{self.table}.{self.column}"
        return self.column


def _clean_expression(expression: str) -> str:
    base = expression.split("|", 1)[0].strip()
    return base


def _parse_field(expression: str) -> FieldReference:
    base = _clean_expression(expression)
    if not base:
        msg = "Encountered empty Jinja placeholder."
        raise ValueError(msg)

    if "." in base:
        table, column = base.split(".", 1)
        table = table.strip() or None
        column = column.strip()
    else:
        table, column = None, base

    if not column:
        msg = f"Invalid field expression: {expression!r}"
        raise ValueError(msg)

    return FieldReference(expression=base, table=table, column=column)


def iter_field_references(template: str) -> Iterator[FieldReference]:
    """Yield field references in the order they appear within *template*."""

    for match in JINJA_PLACEHOLDER.finditer(template):
        yield _parse_field(match.group("expr"))


def extract_field_references(templates: Iterable[str]) -> list[FieldReference]:
    """Extract unique field references from the provided templates preserving order."""

    ordered: "OrderedDict[str, FieldReference]" = OrderedDict()
    for template in templates:
        for reference in iter_field_references(template):
            ordered.setdefault(reference.expression, reference)
    return list(ordered.values())


__all__ = ["FieldReference", "extract_field_references", "iter_field_references"]
