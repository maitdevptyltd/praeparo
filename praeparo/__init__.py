"""Praeparo package public interface."""

from .models.matrix import MatrixConfig, MatrixTotals, MatrixValueConfig

__all__ = ["MatrixConfig", "MatrixTotals", "MatrixValueConfig"]
