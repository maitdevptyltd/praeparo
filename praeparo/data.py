"""Mocked data providers for proof-of-concept pipelines."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

from .models import MatrixConfig
from .templating import FieldReference


@dataclass
class MockResultSet:
    """Synthetic dataset representing the outcome of a faux DAX execution."""

    rows: list[dict[str, object]]
    row_fields: tuple[FieldReference, ...]


def _seed_value(base: int, multiplier: int) -> float:
    return round(base * multiplier / 100.0, 4)


def mock_matrix_data(config: MatrixConfig, row_fields: Iterable[FieldReference]) -> MockResultSet:
    """Generate deterministic sample data for a matrix visual."""

    ordered_fields = tuple(row_fields)
    generated_rows: list[dict[str, object]] = []

    for index in range(1, 4):
        row: dict[str, object] = {}
        for position, field in enumerate(ordered_fields, start=1):
            seed = f"{field.column.replace('_', ' ').title()} {index}"
            row[field.placeholder] = seed
        for value_position, value in enumerate(config.values, start=1):
            multiplier = index * value_position
            if value.format and value.format.startswith("percent"):
                row[value.label or value.id] = _seed_value(multiplier, 5)
            elif value.format and value.format.startswith("duration"):
                row[value.label or value.id] = multiplier * 900
            else:
                row[value.label or value.id] = multiplier * 100
        generated_rows.append(row)

    return MockResultSet(rows=generated_rows, row_fields=ordered_fields)


__all__ = ["MockResultSet", "mock_matrix_data"]
