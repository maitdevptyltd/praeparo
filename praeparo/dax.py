"""Utilities for generating DAX query text from Praeparo models."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Sequence

from .models import MatrixConfig
from .templating import FieldReference

if TYPE_CHECKING:
    from .models import MatrixFilterConfig


@dataclass(frozen=True)
class DaxQueryPlan:
    """Represents the components of a generated DAX query."""

    statement: str
    rows: tuple[FieldReference, ...]
    values: tuple[str, ...]
    define: str | None = None


SHOW_AS_PERCENT_COLUMN_TOTAL = "percent of column total"


def _escape_label(label: str) -> str:
    return label.replace('"', '""')


def _format_measure(identifier: str) -> str:
    trimmed = identifier.strip()
    if trimmed.startswith("[") and trimmed.endswith("]"):
        return trimmed
    return f"[{trimmed}]"


def _apply_show_as(show_as: str | None, measure: str, row_fields: Sequence[FieldReference]) -> str:
    if not show_as:
        return measure

    normalized = show_as.strip().lower()
    if normalized == SHOW_AS_PERCENT_COLUMN_TOTAL and row_fields:
        clauses = []
        primary = row_fields[-1]
        clauses.append(f"REMOVEFILTERS({primary.dax_reference})")
        for field in row_fields[:-1]:
            target = field.table or field.dax_reference
            clauses.append(f"REMOVEFILTERS({target})")
        arguments = ", ".join(clauses)
        return f"DIVIDE({measure}, CALCULATE({measure}, {arguments}))"

    return measure


def _split_field_reference(expression: str) -> tuple[str, str]:
    table, column = expression.split(".", 1)
    table = table.strip()
    column = column.strip()
    return table, column


def _format_column_reference(expression: str) -> tuple[str, str, str]:
    table, column = _split_field_reference(expression)
    reference = f"{table}[{column}]"
    return table, column, reference


def _escape_filter_value(value: str) -> str:
    return value.replace('"', '""')


def _format_filter_clause(filter_config: "MatrixFilterConfig") -> str:
    if filter_config.expression:
        return filter_config.expression
    if not filter_config.field or not filter_config.include:
        msg = "Filter configuration must define either an expression or field/include pair."
        raise ValueError(msg)
    _table, _column, column_reference = _format_column_reference(filter_config.field)
    values = ", ".join(f'"{_escape_filter_value(item)}"' for item in filter_config.include)
    return f"{column_reference} IN {{ {values} }}"


def _summarize_columns(row_lines: Sequence[str], value_lines: Sequence[str]) -> str:
    inner_parts: list[str] = []
    inner_parts.extend(row_lines)
    inner_parts.extend(value_lines)

    inner_body = ",\n    ".join(inner_parts) if inner_parts else ""
    return "SUMMARIZECOLUMNS(\n    " + inner_body + "\n)"


def _indent_block(text: str) -> str:
    lines = text.splitlines()
    if not lines:
        return ""
    return "    " + "\n    ".join(lines)


def _wrap_with_filters(body: str, filter_lines: Sequence[str]) -> str:
    if not filter_lines:
        return body

    indented_body = _indent_block(body)
    filter_block = ",\n".join("    " + line for line in filter_lines)
    return "CALCULATETABLE(\n" + indented_body + ",\n" + filter_block + "\n)"


def build_matrix_query(config: MatrixConfig, row_fields: Sequence[FieldReference]) -> DaxQueryPlan:
    """Construct a simple SUMMARIZECOLUMNS query for the given matrix configuration."""

    ordered_rows = tuple(row_fields)
    row_lines: list[str] = [reference.dax_reference for reference in ordered_rows]

    value_lines: list[str] = []
    measure_names: list[str] = []
    for value in config.values:
        alias = _escape_label(value.label or value.id)
        measure = _format_measure(value.id)
        expression = _apply_show_as(value.show_as, measure, ordered_rows)
        measure_names.append(measure)
        value_lines.append(f'"{alias}", {expression}')

    filter_lines = [_format_filter_clause(filter_config) for filter_config in config.filters]

    summarize = _summarize_columns(row_lines, value_lines)
    body = summarize if not filter_lines else _wrap_with_filters(summarize, filter_lines)

    define_block: str | None = None
    if config.define:
        candidate = config.define.strip()
        if candidate:
            define_block = candidate

    parts: list[str] = []
    if define_block:
        parts.append("DEFINE\n" + define_block)
    parts.append("EVALUATE\n" + body)
    statement = "\n\n".join(parts)

    return DaxQueryPlan(
        statement=statement,
        rows=ordered_rows,
        values=tuple(measure_names),
        define=define_block,
    )


__all__ = ["DaxQueryPlan", "build_matrix_query"]

