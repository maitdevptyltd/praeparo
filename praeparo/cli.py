"""Command line interface for Praeparo proof-of-concept pipelines."""

from __future__ import annotations

import argparse
import asyncio
from pathlib import Path
from typing import Sequence

from .data import MatrixResultSet, mock_matrix_data, powerbi_matrix_data
from .datasources import DataSourceConfigError, ResolvedDataSource, resolve_datasource
from .dax import build_matrix_query
from .io.yaml_loader import ConfigLoadError, load_visual_config
from .models import FrameConfig, MatrixConfig
from .powerbi import (
    PowerBIAuthenticationError,
    PowerBIConfigurationError,
    PowerBIQueryError,
)
from .rendering import frame_html, frame_png, matrix_html, matrix_png
from .templating import extract_field_references


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Render a Praeparo visual from a YAML configuration."
    )
    parser.add_argument("config", type=Path, help="Path to the visual YAML file.")
    parser.add_argument(
        "--out",
        type=Path,
        default=None,
        help="Destination for the generated HTML output (defaults to <project>/build/<name>.html).",
    )
    parser.add_argument(
        "--png-out",
        type=Path,
        default=None,
        help="Optional destination for a static PNG snapshot of the visual.",
    )
    parser.add_argument(
        "--data-source",
        type=str,
        default=None,
        help="Name or path of the data source definition to use (overrides visual configuration).",
    )
    parser.add_argument(
        "--dataset-id",
        type=str,
        default=None,
        help="Execute the DAX query against the specified Power BI dataset instead of mock data.",
    )
    parser.add_argument(
        "--workspace-id",
        type=str,
        default=None,
        help="Optional workspace (group) id when querying a dataset via the Power BI API.",
    )
    parser.add_argument(
        "--print-dax",
        action="store_true",
        help="Print the generated DAX statement(s) to stdout.",
    )
    return parser


def _project_root_for(path: Path) -> Path | None:
    current = path.parent
    while True:
        if current.name == "visuals":
            return current.parent
        if (current / "visuals").is_dir():
            return current
        if current.parent == current:
            return None
        current = current.parent


def _default_output_path(
    config_path: Path, project_root: Path | None, extension: str
) -> Path:
    base = project_root or config_path.parent
    build_dir = base / "build"
    return build_dir / f"{config_path.stem}.{extension}"


def _ensure_directory(path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)


def _resolve_datasource(
    config: MatrixConfig,
    args: argparse.Namespace,
    *,
    visual_path: Path,
) -> ResolvedDataSource:
    reference = args.datasource if args.datasource is not None else config.datasource
    return resolve_datasource(reference, visual_path=visual_path)


def _load_dataset(
    config: MatrixConfig,
    row_fields,
    query,
    args: argparse.Namespace,
    parser: argparse.ArgumentParser,
    *,
    visual_path: Path,
) -> MatrixResultSet | None:
    if args.dataset_id:
        try:
            return asyncio.run(
                powerbi_matrix_data(
                    config,
                    row_fields,
                    query,
                    dataset_id=args.dataset_id,
                    group_id=args.workspace_id,
                )
            )
        except (
            PowerBIConfigurationError,
            PowerBIAuthenticationError,
            PowerBIQueryError,
        ) as exc:
            parser.error(str(exc))
            return None

    try:
        datasource = _resolve_datasource(config, args, visual_path=visual_path)
    except DataSourceConfigError as exc:
        parser.error(str(exc))
        return None

    if datasource.type == "mock":
        return mock_matrix_data(config, row_fields)

    dataset_id = datasource.dataset_id
    if not dataset_id:
        parser.error(
            f"Data source '{datasource.name}' does not define a dataset_id and no --dataset-id override was provided."
        )
        return None

    workspace_id = args.workspace_id or datasource.workspace_id
    settings = datasource.settings

    try:
        return asyncio.run(
            powerbi_matrix_data(
                config,
                row_fields,
                query,
                dataset_id=dataset_id,
                group_id=workspace_id,
                settings=settings,
            )
        )
    except (
        PowerBIConfigurationError,
        PowerBIAuthenticationError,
        PowerBIQueryError,
    ) as exc:
        parser.error(str(exc))
        return None


def _matrix_png(
    config: MatrixConfig,
    dataset: MatrixResultSet,
    path: Path,
    parser: argparse.ArgumentParser,
) -> bool:
    try:
        _ensure_directory(path)
        matrix_png(config, dataset, str(path))
        return True
    except RuntimeError as exc:
        parser.error(str(exc))
        return False


def _frame_png(
    frame: FrameConfig,
    datasets: list[tuple[MatrixConfig, MatrixResultSet]],
    path: Path,
    parser: argparse.ArgumentParser,
) -> bool:
    try:
        _ensure_directory(path)
        frame_png(frame, datasets, str(path))
        return True
    except RuntimeError as exc:
        parser.error(str(exc))
        return False


def _render_matrix(
    config: MatrixConfig,
    args: argparse.Namespace,
    parser: argparse.ArgumentParser,
    *,
    config_path: Path,
    project_root: Path | None,
) -> int:
    row_fields = extract_field_references([row.template for row in config.rows])
    query = build_matrix_query(config, row_fields)

    dataset = _load_dataset(
        config, row_fields, query, args, parser, visual_path=config_path
    )
    if dataset is None:
        return 2

    outputs: list[Path] = []

    out_path = args.out or _default_output_path(config_path, project_root, "html")
    _ensure_directory(out_path)
    matrix_html(config, dataset, str(out_path))
    outputs.append(out_path)

    png_path = args.png_out
    if png_path is not None:
        if not _matrix_png(config, dataset, png_path, parser):
            return 2
        outputs.append(png_path)

    if args.print_dax:
        print(query.statement)

    if outputs:
        rendered = ", ".join(str(path) for path in outputs)
        print(f"Wrote matrix visualization to {rendered}")
    return 0


def _render_frame(
    frame: FrameConfig,
    args: argparse.Namespace,
    parser: argparse.ArgumentParser,
    *,
    config_path: Path,
    project_root: Path | None,
) -> int:
    child_outputs: list[tuple[MatrixConfig, MatrixResultSet]] = []
    printed_dax: list[str] = []

    for child in frame.children:
        config = child.config
        row_fields = extract_field_references([row.template for row in config.rows])
        query = build_matrix_query(config, row_fields)

        dataset = _load_dataset(
            config, row_fields, query, args, parser, visual_path=child.source
        )
        if dataset is None:
            return 2

        child_outputs.append((config, dataset))
        printed_dax.append(f"-- {config.title or child.source.name}\n{query.statement}")

    out_path = args.out or _default_output_path(config_path, project_root, "html")
    _ensure_directory(out_path)
    frame_html(frame, child_outputs, str(out_path))

    outputs = [out_path]
    png_path = args.png_out
    if png_path is not None:
        if not _frame_png(frame, child_outputs, png_path, parser):
            return 2
        outputs.append(png_path)

    if args.print_dax:
        print("\n\n".join(printed_dax))

    rendered = ", ".join(str(path) for path in outputs)
    print(f"Wrote frame visualization to {rendered}")
    return 0


def run(argv: Sequence[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)

    try:
        visual = load_visual_config(args.config)
    except ConfigLoadError as exc:
        parser.error(str(exc))
        return 2

    project_root = _project_root_for(args.config)

    if isinstance(visual, MatrixConfig):
        return _render_matrix(
            visual, args, parser, config_path=args.config, project_root=project_root
        )

    if isinstance(visual, FrameConfig):
        return _render_frame(
            visual, args, parser, config_path=args.config, project_root=project_root
        )

    parser.error(f"Unsupported visual type in {args.config}")
    return 2


def main() -> None:
    raise SystemExit(run())


if __name__ == "__main__":
    main()
