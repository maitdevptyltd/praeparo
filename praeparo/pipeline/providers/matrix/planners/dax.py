"""Matrix query planner backed by a DAX execution client."""

from __future__ import annotations

import asyncio
import inspect
from pathlib import Path
from typing import Awaitable, Sequence, TYPE_CHECKING, Callable

from praeparo.data import MatrixResultSet, mock_matrix_data
from praeparo.dax import DaxQueryPlan, build_matrix_query
from praeparo.datasources import DataSourceConfigError, ResolvedDataSource, resolve_datasource
from praeparo.models import MatrixConfig
from praeparo.templating import FieldReference, extract_field_references

from .base import MatrixPlannerResult, MatrixQueryPlanner
from ...dax.clients.base import DaxExecutionClient

if TYPE_CHECKING:  # pragma: no cover
    from ....core import ExecutionContext


class DaxBackedMatrixPlanner(MatrixQueryPlanner):
    """Plans matrix visuals by delegating execution to a DAX client."""

    def __init__(
        self,
        *,
        dax_client: DaxExecutionClient,
        datasource_resolver: Callable[[str | None, Path], ResolvedDataSource] | None = None,
        mock_provider: Callable[[MatrixConfig, Sequence[FieldReference]], MatrixResultSet] = mock_matrix_data,
    ) -> None:
        self._dax_client = dax_client
        if datasource_resolver is None:
            def _default_resolver(reference: str | None, visual_path: Path) -> ResolvedDataSource:
                return resolve_datasource(reference, visual_path=visual_path)
            self._resolve_datasource = _default_resolver
        else:
            self._resolve_datasource = datasource_resolver
        self._mock_provider = mock_provider

    def plan(self, config: MatrixConfig, *, context: "ExecutionContext") -> MatrixPlannerResult:
        row_fields = tuple(self._extract_row_fields(config))
        plan = build_matrix_query(config, row_fields)

        data_options = context.options.data
        dataset_override = getattr(data_options, "dataset_id", None)
        workspace_override = getattr(data_options, "workspace_id", None)
        provider_key = self._resolve_provider_key(context, data_options)

        if dataset_override:
            dataset = self._execute_with_override(config, row_fields, plan, dataset_override, workspace_override)
            return MatrixPlannerResult(plan=plan, dataset=dataset)

        if provider_key == "mock":
            dataset = self._mock_provider(config, row_fields)
            return MatrixPlannerResult(plan=plan, dataset=dataset)

        dataset = self._execute_from_datasource(config, row_fields, plan, context, data_options)
        return MatrixPlannerResult(plan=plan, dataset=dataset)

    def _extract_row_fields(self, config: MatrixConfig) -> Sequence[FieldReference]:
        return extract_field_references([row.template for row in config.rows])

    def _resolve_provider_key(self, context: "ExecutionContext", data_options) -> str | None:
        case_key = context.case_key
        overrides = getattr(data_options, "provider_case_overrides", {}) or {}
        if case_key and case_key in overrides:
            candidate = overrides[case_key].strip().lower()
            if candidate:
                return candidate
        provider_key = getattr(data_options, "provider_key", None)
        if provider_key:
            candidate = provider_key.strip().lower()
            if candidate:
                return candidate
        return None

    def _execute_with_override(
        self,
        config: MatrixConfig,
        row_fields: Sequence[FieldReference],
        plan: DaxQueryPlan,
        dataset_id: str,
        workspace_override: str | None,
    ) -> MatrixResultSet:
        result = self._dax_client.execute_matrix(
            config,
            row_fields,
            plan,
            dataset_id=dataset_id,
            workspace_id=workspace_override,
        )
        return self._resolve_result(result)

    def _execute_from_datasource(
        self,
        config: MatrixConfig,
        row_fields: Sequence[FieldReference],
        plan: DaxQueryPlan,
        context: "ExecutionContext",
        data_options,
    ) -> MatrixResultSet:
        reference = getattr(data_options, "datasource_override", None) or config.datasource
        visual_path = context.config_path
        if visual_path is None:
            msg = "Matrix execution requires a config_path to resolve datasources."
            raise DataSourceConfigError(msg)

        datasource = self._resolve_datasource(reference, visual_path)
        if datasource.type == "mock":
            return self._mock_provider(config, row_fields)

        dataset_id = datasource.dataset_id
        if not dataset_id:
            title = config.title or "matrix"
            msg = f"Data source '{datasource.name}' for {title} lacks a dataset_id."
            raise DataSourceConfigError(msg)

        workspace_override = getattr(data_options, "workspace_id", None)
        result = self._dax_client.execute_matrix(
            config,
            row_fields,
            plan,
            dataset_id=dataset_id,
            workspace_id=workspace_override or datasource.workspace_id,
            settings=datasource.settings,
        )
        return self._resolve_result(result)

    def _resolve_result(
        self,
        result: MatrixResultSet | Awaitable[MatrixResultSet],
    ) -> MatrixResultSet:
        if inspect.isawaitable(result):
            try:
                resolved = asyncio.run(result)  # type: ignore[arg-type]
            except RuntimeError as exc:  # pragma: no cover
                msg = (
                    "DAX execution returned an awaitable while an event loop is already running. "
                    "Provide a synchronous client or adapt the call site to handle async planners."
                )
                raise RuntimeError(msg) from exc
        else:
            resolved = result
        if not isinstance(resolved, MatrixResultSet):
            msg = "DAX execution client must return a MatrixResultSet."
            raise TypeError(msg)
        return resolved

