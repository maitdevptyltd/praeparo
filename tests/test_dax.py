from praeparo.dax import build_matrix_query
from praeparo.models import MatrixConfig, MatrixFilterConfig, MatrixTotals, MatrixValueConfig
from praeparo.templating import FieldReference


def test_build_matrix_query_constructs_summarizecolumns(snapshot) -> None:
    config = MatrixConfig(
        type="matrix",
        rows=["{{dim.City}}"],
        values=[MatrixValueConfig(id="Total Sales", label="Sales")],
        totals=MatrixTotals.OFF,
    )

    row_fields = [FieldReference(expression="dim.City", table="dim", column="City")]

    plan = build_matrix_query(config, row_fields)

    assert plan.statement == snapshot
    assert plan.define is None


def test_build_matrix_query_includes_filters(snapshot) -> None:
    config = MatrixConfig(
        type="matrix",
        rows=["{{dim.City}}"],
        values=[MatrixValueConfig(id="Total Sales", label="Sales")],
        filters=[MatrixFilterConfig(field="dim.City", include=["Seattle", "Portland"])],
    )

    row_fields = [FieldReference(expression="dim.City", table="dim", column="City")]

    plan = build_matrix_query(config, row_fields)

    assert plan.statement == snapshot
    assert plan.define is None




def test_build_matrix_query_supports_expression_filter(snapshot) -> None:
    config = MatrixConfig(
        type="matrix",
        rows=["{{dim.City}}"],
        values=[MatrixValueConfig(id="Total Sales", label="Sales")],
        filters=[MatrixFilterConfig(expression='dim[City] <> "Unknown"')],
    )

    row_fields = [FieldReference(expression="dim.City", table="dim", column="City")]

    plan = build_matrix_query(config, row_fields)

    assert plan.statement == snapshot
    assert plan.define is None


def test_build_matrix_query_with_define(snapshot) -> None:
    config = MatrixConfig(
        type="matrix",
        define="MEASURE Sales[Total Sales] = SUM(Sales[Amount])",
        rows=["{{Sales.Region}}"],
        values=[MatrixValueConfig(id="Total Sales", label="Total Sales")],
    )

    row_fields = [FieldReference(expression="Sales.Region", table="Sales", column="Region")]

    plan = build_matrix_query(config, row_fields)

    assert plan.statement == snapshot
    assert plan.define == "MEASURE Sales[Total Sales] = SUM(Sales[Amount])"

