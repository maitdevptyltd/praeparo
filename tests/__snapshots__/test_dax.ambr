# serializer version: 1
# name: test_build_matrix_query_constructs_summarizecolumns
  '''
  EVALUATE
  SUMMARIZECOLUMNS(
      dim[City],
      "Sales", [Total Sales]
  )
  '''
# ---
# name: test_build_matrix_query_includes_filters
  '''
  EVALUATE
  CALCULATETABLE(
      SUMMARIZECOLUMNS(
          dim[City],
          "Sales", [Total Sales]
      ),
      dim[City] IN { "Seattle", "Portland" }
  )
  '''
# ---
# name: test_build_matrix_query_supports_expression_filter
  '''
  EVALUATE
  CALCULATETABLE(
      SUMMARIZECOLUMNS(
          dim[City],
          "Sales", [Total Sales]
      ),
      dim[City] <> "Unknown"
  )
  '''
# ---
# name: test_build_matrix_query_with_define
  '''
  DEFINE
  MEASURE Sales[Total Sales] = SUM(Sales[Amount])
  
  EVALUATE
  SUMMARIZECOLUMNS(
      Sales[Region],
      "Total Sales", [Total Sales]
  )
  '''
# ---
